sbat.c

```diff
@@ -335,4 +335,75 @@ parse_sbat_var(list_t *entries)
 	FreePool(data);
 	return EFI_INVALID_PARAMETER;
 }
+
+EFI_STATUS
+set_sbat_uefi_variable(void)
+{
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	UINT32 attributes = 0;
+
+	CHAR8 *sbat = NULL;
+	UINTN sbatsize = 0;
+
+	/* get size */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+	sbat = AllocateZeroPool(sbatsize);
+	/* read and get attributes */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+
+	/*
+	 * If attributes and the initial part of the version match, then
+	 * there is no need to re-write it. If it is not set or the read
+	 * failed, then attributes won't match.
+	 *
+	 * Finally, if the SBAT_VAR built into this shim is newer, update
+	 * it. Either the sbat version or date can be newer, and since
+	 * the version is more significant than the date, strcmp() applies.
+	 */
+	if (((attributes != (EFI_VARIABLE_NON_VOLATILE |
+		EFI_VARIABLE_BOOTSERVICE_ACCESS))
+		|| (attributes != (EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)))
+		|| (strncmp(sbat, SBAT_VAR_SIG, strlen(SBAT_VAR_SIG)) != 0)
+		|| (strncmp(SBAT_VAR, sbat, strlen(SBAT_VAR)) > 0)
+		) {
+
+		/* delete previous variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				attributes, 0, "");
```

Comment:
I think we want to check returned EFI status here and report if failed to set.

Reply:
I'm adding dprintfs only. UEFI failures should not prevent use from ultimately trying to set the variable since that may in fact succeed and is unlikely to do anything worse than not setting it.